daemon/c_seccomp/ioctl.c

```diff
@@ -0,0 +1,260 @@
+
+/*
+ * This partial filter continues all system calls and only handles
+ * (notify) SYS_ioctl.
+ * Remember that this needs to be integrated in the overall filter
+ * since we assume syscall nr already loaded and if it is not
+ * SYS_ioctl, we jump one instruction beyond the last instruction
+ * of filter_ioctl to the next comparison of syscall number.
+ */
+static struct sock_filter c_seccomp_filter_ioctl[] = {
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, SYS_ioctl, 0, 6),
+	BPF_STMT(BPF_LD | BPF_W | BPF_ABS, (offsetof(struct seccomp_data, args[1]))),
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, RTC_EPOCH_SET, 3, 0),
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, RTC_SET_TIME, 2, 0),
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, RTC_PARAM_SET, 1, 0),
+
+	BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
+	BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_USER_NOTIF),
+};
+
+static int
+rtc_ioctl(unsigned int fd, unsigned int cmd, unsigned long param)
+{
+	return syscall(__NR_ioctl, fd, cmd, param);
+}
+
+struct ioctl_fork_data {
+	unsigned int fd;
+	unsigned int cmd;
+	unsigned long param;
+};
+
+static int
+c_seccomp_do_ioctl_fork(const void *data)
+{
+	const struct ioctl_fork_data *params = data;
+	ASSERT(params);
+
+	DEBUG("Executing ioctl in namespaces of container");
+	int ret = rtc_ioctl(params->fd, params->cmd, params->param);
+	if (ret == -1)
+		ERROR_ERRNO("rtc_ioctl() failed in namespaces of container!");
+
+	return ret;
+}
+
+static int
+c_seccomp_proc_get_rtc_major(void)
+{
+	FILE *file = fopen("/proc/devices", "r");
+	char line[256];
+	int major = -1;
+	int n = -1;
+
+	IF_NULL_RETVAL(file, -1);
+
+	while (fgets(line, sizeof(line), file)) {
+		if (!strstr(line, "rtc"))
+			continue;
+
+		n = sscanf(line, "%d rtc", &major);
+		TRACE("line: %s, n=%d; major=%d", line, n, major);
+		break;
+	}
+
+	fclose(file);
+
+	IF_TRUE_RETVAL(n != 1, -1);
+
+	TRACE("Parsed major of /dev/rtc*: %d", major);
+	return major;
+}
+
+struct sock_filter *
+c_seccomp_ioctl_get_filter(c_seccomp_t *seccomp, int *size)
+{
+	*size = 0;
+
+	if ((!(COMPARTMENT_FLAG_SYSTEM_TIME & compartment_get_flags(seccomp->compartment)) ||
+	     c_seccomp_proc_get_rtc_major() <= 0)) {
+		DEBUG("Compartment seccomp filter will not handle ioctl() syscalls");
+		return NULL;
+	}
+
+	*size = sizeof(c_seccomp_filter_ioctl);
+
+	INFO("Enable rtc specific ioctl() emulation!");
+	return c_seccomp_filter_ioctl;
+}
+
+int
+c_seccomp_emulate_ioctl(c_seccomp_t *seccomp, struct seccomp_notif *req,
+			struct seccomp_notif_resp *resp)
+{
+	int ret_ioctl = 0;
+	int fd_in_target = -1;
+	unsigned int cmd = 0;
+	unsigned long param = 0;
+
+	/*
+	 * in any case of error just continue the syscall in the kernel,
+	 * we are only interested in /dev/rtc* devices to allow setting the hw clock
+	 */
+	resp->error = 0;
+	resp->val = 0;
+	resp->flags = SECCOMP_USER_NOTIF_FLAG_CONTINUE;
+
+	if (!(COMPARTMENT_FLAG_SYSTEM_TIME & compartment_get_flags(seccomp->compartment))) {
+		TRACE("Flag SYSTEM_TIME not continue syscall as is: %d", req->data.nr);
```

Comment:
Flag SYSTEM_TIME not <set>. Continue syscall as is:...

Reply:
done